File: src/kotoba/lang/kami_nv_compat/actions/action_term.clj
(ns kotoba.lang.kami-nv-compat.actions.action-term
  "ActionTerm base + ActionManager — JVM port of src/actions/action-term.ts.
  Mirrors isaaclab.envs.mdp.actions.ActionTerm (Isaac Lab 1.x). One ActionTerm
  processes one slice of the policy's flat action vector (process-actions! →
  scale + offset), then writes the result onto the env (apply-actions!).
  ActionManager composes multiple terms into a single combined action vector.
  Wave 16c of ADR-2607020130."
  (:require [kotoba.lang.kami-nv-compat.actions.articulated-env :as ae]))

(defprotocol IActionTerm
  (action-dim       [this])
  (process-actions! [this raw])
  (apply-actions!   [this env])
  (reset-term!      [this]))

(defn make-action-term-state
  "Build the base state map for an ActionTerm: cfg, action-dim, raw/processed
  action atoms. Subclasses call this then reify IActionTerm."
  [cfg]
  (when (or (nil? (:joint-names cfg)) (empty? (:joint-names cfg)))
    (throw (ex-info "ActionTerm: cfg.joint-names must be non-empty" {:cfg cfg})))
  (let [dim (or (:action-dim cfg) (count (:joint-names cfg)))]
    {:cfg cfg
     :action-dim dim
     :raw-actions (atom (vec (repeat dim 0)))
     :processed-actions (atom (vec (repeat dim 0)))}))

(defn base-process-actions!
  "Default process: scale + offset element-wise into processed-actions."
  [term raw]
  (let [dim (:action-dim term)
        s (get-in term [:cfg :scale] 1.0)
        o (get-in term [:cfg :offset] 0.0)]
    (when (not= (count raw) dim)
      (throw (ex-info (str "ActionTerm: expected " dim " action elements, got " (count raw))
                      {:expected dim :got (count raw)})))
    (reset! (:raw-actions term) (vec raw))
    (reset! (:processed-actions term) (vec (map #(+ (* % s) o) raw)))))

(defn base-reset! [term]
  (let [dim (:action-dim term)]
    (reset! (:raw-actions term) (vec (repeat dim 0)))
    (reset! (:processed-actions term) (vec (repeat dim 0)))))

;; ── ActionManager ─────────────────────────────────────────────────────────

(defn action-manager
  "Build an ActionManager from a vector of IActionTerm instances."
  [terms]
  (when (or (nil? terms) (empty? terms))
    (throw (ex-info "ActionManager requires at least one ActionTerm" {})))
  (let [offsets (loop [ts (seq terms) off 0 acc []]
                  (if (empty? ts)
                    acc
                    (let [t (first ts)]
                      (recur (rest ts) (+ off (action-dim t)) (conj acc off)))))]
    {:terms terms
     :offsets offsets
     :total-action-dim (reduce + (map action-dim terms))}))

(defn manager-process-actions! [mgr raw]
  (let [expected (:total-action-dim mgr)]
    (when (not= (count raw) expected)
      (throw (ex-info (str "ActionManager: expected " expected " action elements, got " (count raw))
                      {:expected expected :got (count raw)})))
    (doseq [[i term] (map-indexed vector (:terms mgr))]
      (let [start (nth (:offsets mgr) i)]
        (process-actions! term (subvec (vec raw) start (+ start (action-dim term))))))))

(defn manager-apply-actions! [mgr env]
  (doseq [term (:terms mgr)] (apply-actions! term env)))

(defn manager-reset! [mgr]
  (doseq [term (:terms mgr)] (reset-term! term)))

File: src/actions/action-term.ts
// ActionTerm base + ActionTermCfgBase.
//
// Mirror of `isaaclab.envs.mdp.actions.ActionTerm` (Isaac Lab 1.x). One
// ActionTerm processes one slice of the policy's flat action vector
// (`process_actions(raw)` → `processed_actions` via scale + offset),
// then writes the result onto the env (`apply_actions(env)`).
//
// `ActionManager` composes multiple ActionTerm instances into a single
// combined action vector — slices `raw` by per-term offsets and
// dispatches each slice.

import type { ArticulatedEnv } from "./articulated-env.js";

export interface ActionTermCfgBase {
  assetName?: string;
  /** Integer joint indices (env-specific) the action term targets. */
  jointNames: number[];
  /** Multiplied element-wise with the raw action. Default 1.0. */
  scale?: number;
  /** Added element-wise after scale. Default 0.0. */
  offset?: number;
  /** Optional override; if undefined, inferred from jointNames length. */
  actionDim?: number;
}

export abstract class ActionTerm {
  cfg: ActionTermCfgBase;
  readonly actionDim: number;
  raw_actions: number[];
  processed_actions: number[];

  constructor(cfg: ActionTermCfgBase) {
    if (!cfg.jointNames || cfg.jointNames.length === 0) {
      throw new Error(`${this.constructor.name}.cfg.jointNames must be non-empty`);
    }
    this.cfg = cfg;
    this.actionDim = cfg.actionDim ?? cfg.jointNames.length;
    this.raw_actions = new Array<number>(this.actionDim).fill(0);
    this.processed_actions = new Array<number>(this.actionDim).fill(0);
  }

  /** Default impl: scale + offset element-wise into processed_actions. */
  processActions(raw: readonly number[]): void {
    if (raw.length !== this.actionDim) {
      throw new Error(
        `${this.constructor.name}: expected ${this.actionDim} action elements, got ${raw.length}`,
      );
    }
    this.raw_actions = [...raw];
    const s = this.cfg.scale ?? 1.0;
    const o = this.cfg.offset ?? 0.0;
    this.processed_actions = raw.map((r) => r * s + o);
  }

  abstract applyActions(env: ArticulatedEnv): void;

  reset(): void {
    this.raw_actions = new Array<number>(this.actionDim).fill(0);
    this.processed_actions = new Array<number>(this.actionDim).fill(0);
  }
}

// ── ActionManager ────────────────────────────────────────────────────────

export class ActionManager {
  readonly terms: ActionTerm[];
  readonly termNames: string[];
  readonly totalActionDim: number;
  private readonly _offsets: number[];

  constructor(terms: ActionTerm[]) {
    if (!terms || terms.length === 0) {
      throw new Error("ActionManager requires at least one ActionTerm");
    }
    this.terms = terms;
    this.termNames = terms.map((t) => t.constructor.name);
    this._offsets = [];
    let off = 0;
    for (const t of terms) {
      this._offsets.push(off);
      off += t.actionDim;
    }
    this.totalActionDim = off;
  }

  /** Slice `raw` and dispatch to each term in order. */
  processActions(raw: readonly number[]): void {
    if (raw.length !== this.totalActionDim) {
      throw new Error(
        `ActionManager: expected ${this.totalActionDim} action elements, got ${raw.length}`,
      );
    }
    for (let i = 0; i < this.terms.length; i++) {
      const start = this._offsets[i];
      const end = start + this.terms[i].actionDim;
      this.terms[i].processActions(raw.slice(start, end));
    }
  }

  applyActions(env: ArticulatedEnv): void {
    for (const t of this.terms) t.applyActions(env);
  }

  reset(_envIds?: readonly number[]): void {
    for (const t of this.terms) t.reset();
  }

  numTerms(): number {
    return this.terms.length;
  }
}
